File: packages/territory-rules/src/reckoning.ts
/**
 * Weekly reckoning for one cell (`docs/territory-rules.md` "Weekly reckoning",
 * `plan.md` "Shared Rule Semantics" item 9). Pure: idempotency per week is the job's concern.
 */
import { RULES } from './config.js';
import type { FactionStrength, ReckonInput, ReckonResult } from './types.js';

/** Strengths below this are dropped from the result. */
export const MIN_TRACKED_STRENGTH = 0.001;

/**
 * `strength' = strength * DECAY + sum(cappedMeters) + sum(bonus meters)` per faction, then the
 * ownership table (all comparisons use the new strengths):
 *
 * | Situation | Owner |
 * |---|---|
 * | no faction >= MIN_STRENGTH_M | unclaimed |
 * | exact tie at the top (among factions >= MIN) | incumbent if it is still >= MIN, else unclaimed |
 * | no incumbent, or the incumbent leads | the leader |
 * | incumbent < MIN (it no longer holds the cell) | the leader; hysteresis does not apply |
 * | challenger >= incumbent * (1 + HYSTERESIS) | the challenger |
 * | otherwise | incumbent keeps |
 *
 * `captain` is the owning faction's top contributor by `cappedMeters` among contributions that
 * carry a `userId` (ties by `userId` ascending); bonuses do not count.
 */
export function reckonWeek(input: ReckonInput): ReckonResult {
  const totals = new Map<number, number>();
  const add = (factionId: number, value: number): void => {
    totals.set(factionId, (totals.get(factionId) ?? 0) + value);
  };
  for (const s of input.strengths) add(s.factionId, s.strength * RULES.DECAY);
  for (const c of input.contributions) add(c.factionId, c.cappedMeters);
  for (const b of input.bonuses) add(b.factionId, b.meters);

  const strengths: FactionStrength[] = [...totals.entries()]
    .filter(([, strength]) => strength >= MIN_TRACKED_STRENGTH)
    .map(([factionId, strength]) => ({ factionId, strength }))
    .sort((a, b) => a.factionId - b.factionId);

  const strengthOf = (factionId: number | null): number =>
    factionId === null ? 0 : (strengths.find((s) => s.factionId === factionId)?.strength ?? 0);
  const incumbent = input.owner;
  const incumbentHolds = incumbent !== null && strengthOf(incumbent) >= RULES.MIN_STRENGTH_M;

  const eligible = strengths.filter((s) => s.strength >= RULES.MIN_STRENGTH_M);
  let owner: number | null;
  if (eligible.length === 0) {
    owner = null;
  } else {
    let top = eligible[0]!.strength;
    for (const s of eligible) if (s.strength > top) top = s.strength;
    const leaders = eligible.filter((s) => s.strength === top);
    if (leaders.length > 1) {
      owner = incumbentHolds ? incumbent : null;
    } else {
      const leader = leaders[0]!.factionId;
      if (incumbent === null || leader === incumbent || !incumbentHolds) owner = leader;
      else if (top >= strengthOf(incumbent) * (1 + RULES.HYSTERESIS)) owner = leader;
      else owner = incumbent;
    }
  }

  const flipped = owner !== incumbent;
  let captain: string | null = null;
  let captainMeters = 0;
  if (owner !== null) {
    for (const c of input.contributions) {
      if (c.factionId !== owner || c.userId === undefined || c.cappedMeters <= 0) continue;
      if (
        captain === null ||
        c.cappedMeters > captainMeters ||
        (c.cappedMeters === captainMeters && c.userId < captain)
      ) {
        captain = c.userId;
        captainMeters = c.cappedMeters;
      }
    }
  }

  const result: ReckonResult = { strengths, owner, flipped, captain };
  if (flipped) result.event = { from: incumbent, to: owner };
  return result;
}

File: apps/ios/Packages/TerritoryRules/Sources/TerritoryRules/Reckoning.swift
import Foundation
import H3Kit

/// One cell's weekly reckoning — exactly `docs/territory-rules.md` "Weekly reckoning" (plan.md item 9).
///
/// `strength' = strength × DECAY + Σ cappedMeters + Σ bonusMeters` per faction; factions whose new strength is
/// below `Rules.minRetainedStrengthM` are dropped. Ownership follows the table: nobody ≥ `MIN_STRENGTH_M` →
/// unclaimed; no incumbent (or an incumbent that fell below the minimum) → the unique strongest faction ≥ minimum,
/// unclaimed on a tie; an incumbent ≥ minimum keeps the cell unless a single challenger reaches
/// `incumbent × (1 + HYSTERESIS)`; an exact tie at the top keeps the incumbent. Pure: idempotency is the job's concern.
public func reckonWeek(_ input: ReckonInput) -> ReckonResult {
    var strengths: [Int: Double] = [:]
    for strength in input.strengths {
        strengths[strength.factionId, default: 0] += strength.strength * Rules.decay
    }
    for contribution in input.contributions {
        strengths[contribution.factionId, default: 0] += contribution.cappedMeters
    }
    for bonus in input.bonuses {
        strengths[bonus.factionId, default: 0] += bonus.meters
    }
    strengths = strengths.filter { $0.value >= Rules.minRetainedStrengthM }

    let owner = decideOwner(incumbent: input.owner, strengths: strengths)
    let flipped = owner != input.owner
    return ReckonResult(
        strengths: strengths
            .map { FactionStrength(factionId: $0.key, strength: $0.value) }
            .sorted { $0.factionId < $1.factionId },
        owner: owner,
        flipped: flipped,
        event: flipped ? OwnershipEvent(from: input.owner, to: owner) : nil
    )
}

private func decideOwner(incumbent: Int?, strengths: [Int: Double]) -> Int? {
    let eligible = strengths.filter { $0.value >= Rules.minStrengthM }
    guard !eligible.isEmpty else { return nil }

    // An incumbent that fell below the minimum has lost its claim and is treated as absent.
    let incumbentStrength = incumbent.flatMap { eligible[$0] }
    guard let incumbent, let incumbentStrength else {
        return uniqueLeader(among: eligible)
    }

    let threshold = incumbentStrength * (1 + Rules.hysteresis)
    let clearing = eligible.filter { $0.key != incumbent && $0.value >= threshold }
    guard !clearing.isEmpty else { return incumbent }
    // A single strongest challenger takes the cell; challengers tied at the top leave the incumbent in place.
    return uniqueLeader(among: clearing) ?? incumbent
}

/// The faction with the strictly highest strength, or `nil` on an exact tie (or when empty).
private func uniqueLeader(among strengths: [Int: Double]) -> Int? {
    guard let top = strengths.values.max() else { return nil }
    let leaders = strengths.filter { $0.value == top }.keys
    return leaders.count == 1 ? leaders.first : nil
}
